### ISSUE TITLE
PCA tests failing with Dask 2020.12.0

### ISSUE THREAD

**Issue by tomwhite** [pystatgen/sgkit#421]
There were some changes to SVD in the latest release of Dask (https://docs.dask.org/en/latest/changelog.html).

Example failure

```
________ test_pca__stability[stability_test_result0-randomized-chunks0] ________

stability_test_result = ((100, 50), <xarray.Dataset>
Dimensions:                              (components: 2, samples: 50, variants: 100)
Dime...   sample_pca_loading                   (variants, components) float32 dask.array<chunksize=(100, 2), meta=np.ndarray>)
chunks = (-1, -1), algorithm = 'randomized'

    @pytest.mark.parametrize("chunks", [(-1, -1), (25, 25)])
    @pytest.mark.parametrize("algorithm", ["tsqr", "randomized"])
    def test_pca__stability(stability_test_result, chunks, algorithm):
        # Ensure that results are stable across algorithms and that sign flips
        # do not occur when chunking changes
        if algorithm == "tsqr" and all(c > 0 for c in chunks):
            return
        shape, expected = stability_test_result
        ds = simulate_dataset(*shape, chunks=chunks, n_cohort=3)  # type: ignore[misc]
        actual = pca.pca(
            ds, n_components=2, algorithm=algorithm, n_iter=6, random_state=0, merge=False
        )
        # Results are expected to change slightly with chunking, but they
        # will change drastically (far more than 1e-5) if a sign flip occurs
>       xr.testing.assert_allclose(expected, actual, atol=1e-5)  # type: ignore[no-untyped-call]
E       AssertionError: Left and right Dataset objects are not close
E       
E       
E       Differing data variables:
E       L   sample_pca_projection                (samples, components) float32 dask.array<chunksize=(50, 2), meta=np.ndarray>
E       R   sample_pca_projection                (samples, components) float32 dask.array<chunksize=(50, 2), meta=np.ndarray>
E       L   sample_pca_explained_variance_ratio  (components) float32 dask.array<chunksize=(2,), meta=np.ndarray>
E       R   sample_pca_explained_variance_ratio  (components) float32 dask.array<chunksize=(2,), meta=np.ndarray>
E       L   sample_pca_explained_variance        (components) float32 dask.array<chunksize=(2,), meta=np.ndarray>
E       R   sample_pca_explained_variance        (components) float32 dask.array<chunksize=(2,), meta=np.ndarray>
E       L   sample_pca_component                 (variants, components) float32 dask.array<chunksize=(100, 2), meta=np.ndarray>
E       R   sample_pca_component                 (variants, components) float32 dask.array<chunksize=(100, 2), meta=np.ndarray>
E       L   sample_pca_loading                   (variants, components) float32 dask.array<chunksize=(100, 2), meta=np.ndarray>
E       R   sample_pca_loading                   (variants, components) float32 dask.array<chunksize=(100, 2), meta=np.ndarray>
```

**Comment by tomwhite** [comment]
I had a look into this, and the changes in https://github.com/dask/dask/pull/6813 are causing the test failures. It may be possible to get the old behaviour by setting the new `iterator` parameter in `svd_compressed` to `"iterator"`, but it doesn't look like dask-ml has been updated to allow this to be set on `TruncatedSVD`.

However, perhaps there is a better/more correct way of fixing this (cf https://github.com/dask/dask/issues/6757). What do you think @eric-czech?

In the meantime it might be best to pin the Dask version so we can fix the build and not hold up other changes.

BTW there is another fix needed for `test_pca__randomized_allel_comparison` failures:

```
diff --git a/sgkit/tests/test_pca.py b/sgkit/tests/test_pca.py
index 2039381..18eba30 100644
--- a/sgkit/tests/test_pca.py
+++ b/sgkit/tests/test_pca.py
@@ -245,7 +245,7 @@ def test_pca__randomized_allel_comparison(shape, chunks, n_components):
         iterated_power=5,
         random_state=0,
     )
-    assert ds_sg["sample_pca_projection"].values.dtype == np.float64
+    assert ds_sg["sample_pca_projection"].values.dtype == np.float32
     assert ds_sk["sample_pca_projection"].values.dtype == np.float32
     validate_allel_comparison(ds_sg, ds_sk)
```

**Comment by eric-czech** [comment]
> What do you think @eric-czech?

More iterations were necessary to make that test work but that parameter no longer has any effect without also passing `iterator="power"`, which can be done after https://github.com/dask/dask-ml/issues/766.

> In the meantime it might be best to pin the Dask version so we can fix the build and not hold up other changes.

👍 

> BTW there is another fix needed for test_pca__randomized_allel_comparison failures

Ah right, I believe that's expected now after https://github.com/dask/dask/pull/6802.

**Comment by tomwhite** [comment]
I opened https://github.com/dask/dask-ml/pull/767 to fix the dask-ml side and #430 for sgkit (depends on a dask-ml release).

**Comment by eric-czech** [comment]
I apologize for stepping on your toes @tomwhite with https://github.com/dask/dask/issues/7004, which conflicts w/ https://github.com/dask/dask-ml/pull/767.  Now that https://github.com/dask/dask/issues/7004 is in though, I think this should go away after the next dask-ml release (if it also includes a dask release).

**Comment by tomwhite** [comment]
No problem - thanks for getting this fixed in Dask!

I'll leave this open until the next dask-ml release is out and we can verify that it's fixed.

**Comment by tomwhite** [comment]
Fixed in #430
